text/0000-offline-buildpackages.md

```diff
@@ -0,0 +1,455 @@
+```
+
+
+Asset cache creation should:
+  - filter excluded assets specified in the `asset-config` file.
+  - Transform all remaining assets to an image layer filesystem changeset where the asset is provided at `/cnb/assets/<artifact-digest>`.
+  - Order all assets layers diffID.
+  - Add the `io.buildpacks.asset.layers` and `io.buildpacks.asset.metadata` label metadata to the asset image. If name is specified for `io.buildpacks.asset.metadata`, generate one.
+  - set the created time in image config to a constant value.
+  - set the modification time of all non-asset files to a constant value.
+
+#### UX in practice
+Likely the creation of asset packages will follow two different routes for implementation buildpacks and metabuildpackages. This dichotomy follows the existing differences between packaging metabuildpackages, and implementation buildpackages.
+
+##### Implementation Buildpacks
+
+running `pack package-asset <implementation-buildpack.toml>` should produce an asset cache, as a `buildpack.toml` file contains all needed `assets`.
+
+##### Metabuildpacks
+
+running `pack package-asset <metabuildpack-package.toml>` will use a `package.toml` file as this command needs to reference other `buildpackages` images in the `dependencies` array by specifying a `uri` or `image`.
+
+## Metabuildpackage Creation
+
+### Metadata
+
+When creating a metabuildpack, we want to retain all of the asset information from the child buildpacks.
+
+As such `assets` array in `io.buildpacks.buildpack.layers` `io.buildpacks.buildpack.metadata` of the resulting metabuildpack will be a sum of all of the `asset` arrays in its children.
+
+If there are conflicts when merging these entries, the platform may decide to continue or fail.
+
+
+## Builder Creation
+
+When creating a builder we can now specify `asset-cache` packages we wish to include in the builder image. We may also pass an `--asset-config asset.toml` file to excluded particular assets that would normally be included. Additionally there should be the option to include all assets from specified buildpackages using the `--offline` flag (the resulting builders will be very large).
+
+## Using Asset Caches
+
+Asset caches can be added to a build by two mechanisms
+1) specify an asset cache(s) using the `--asset-cache` flag(s) in a `pack build <...>`.
+2) Assets cache layers can be added to a builder image during its creation. These assets will then be available to all builds that use this builder.
+
+
+When asset caches are added to a build/builder we need to verify the following:
+  - The asset is not excluded through a `asset.toml` file passed to `pack build`.
+  - If two assets provide the same `/cnb/assets/<asset-digest>` these two files must have identical contents. Decisions about rewriting these layers to optimize space are left to the platform.
+
+When creating a builder with asset caches:
+  - Asset cache layers should be the final k layers in the builder. These should be ordered by `diff-ID`.
```

Comment:
It probably makes sense to implement this such that the asset layers go last, but maybe we should leave it out of the RFC because it's an implementation detail that doesn't matter to users and might cause performance problems if someone was extending an existing builder at build time or with some future pack feature.
```suggestion
```